src/transpile/fusion.hpp

```diff
@@ -228,83 +222,208 @@ bool Fusion::can_apply_fusion(const op_t& op) const {
+
+      //  3. allocate op.u to u based on u(i, i) and delta
+      u(i          , (i + 0)    ) = mat(0, 0);
+      u(i          , (i + delta)) = mat(0, 1);
+      u((i + delta), (i + 0)    ) = mat(1, 0);
+      u((i + delta), (i + delta)) = mat(1, 1);
+      filled[i] = filled[i + delta] = true;
+    }
+  } else if (src_qubits.size() == 2) { //2-qubit operation
+
+    reg_t sorted_src_qubits = src_qubits;
+    std::sort(sorted_src_qubits.begin(), sorted_src_qubits.end());
+    const cmatrix_t sorted_mat = sort_matrix(src_qubits, sorted_src_qubits, mat);
+
+    // 1. identify low and high delta
+    auto low = std::distance(dst_sorted_qubits.begin(),
+                               std::find(dst_sorted_qubits.begin(), dst_sorted_qubits.end(), sorted_src_qubits[0]));
+
+    auto high = std::distance(dst_sorted_qubits.begin(),
+                                std::find(dst_sorted_qubits.begin(), dst_sorted_qubits.end(), sorted_src_qubits[1]));
+
+    auto low_delta = 1UL << low;
+    auto high_delta = 1UL << high;
+
+    // 2. find op.u(0, 0) position in U
+    for (uint_t i = 0; i < dst_dim; ++i) {
+      if (filled[i])
+        continue;
+
+      //  3. allocate vmat to u based on u(i, i) and delta
+      u(i                           , (i + 0)                     ) = sorted_mat(0, 0);
+      u(i                           , (i + low_delta)             ) = sorted_mat(0, 1);
+      u(i                           , (i + high_delta)            ) = sorted_mat(0, 2);
+      u(i                           , (i + low_delta + high_delta)) = sorted_mat(0, 3);
+      u((i + low_delta)             , (i + 0)                     ) = sorted_mat(1, 0);
+      u((i + low_delta)             , (i + low_delta)             ) = sorted_mat(1, 1);
+      u((i + low_delta)             , (i + high_delta)            ) = sorted_mat(1, 2);
+      u((i + low_delta)             , (i + low_delta + high_delta)) = sorted_mat(1, 3);
+      u((i + high_delta)            , (i + 0)                     ) = sorted_mat(2, 0);
+      u((i + high_delta)            , (i + low_delta)             ) = sorted_mat(2, 1);
+      u((i + high_delta)            , (i + high_delta)            ) = sorted_mat(2, 2);
+      u((i + high_delta)            , (i + low_delta + high_delta)) = sorted_mat(2, 3);
+      u((i + low_delta + high_delta), (i + 0)                     ) = sorted_mat(3, 0);
+      u((i + low_delta + high_delta), (i + low_delta)             ) = sorted_mat(3, 1);
+      u((i + low_delta + high_delta), (i + high_delta)            ) = sorted_mat(3, 2);
+      u((i + low_delta + high_delta), (i + low_delta + high_delta)) = sorted_mat(3, 3);
+
+      filled[i] = true;
+      filled[i + low_delta] = true;
+      filled[i + high_delta] = true;
+      filled[i + low_delta + high_delta] = true;
+    }
+    //TODO: } else if (src_qubits.size() == 3) {
+  } else {
+    std::stringstream ss;
+    ss << "Fusion::illegal qubit number:" << src_qubits.size();
+    throw std::runtime_error(ss.str());
```

Comment:
We want to fail as fast as possible, so I'd move this `else` to the begginning of the function and write something like:
```c++
cmatrix_t Fusion::expand_matrix(const reg_t& src_qubits, const reg_t& dst_sorted_qubits, const cmatrix_t& mat) const {

if( src_qubits.size() > 2 )
    throw std::runtime_error("Fusion::illegal qubit number: " + std::to_string(src_qubits.size()));
```
There's no need for a `stringstream`.